pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -1200,12 +1219,45 @@ public void trackSegmentForUpsertView(IndexSegment segment) {
     if (_upsertViewManager != null) {
       _upsertViewManager.trackSegment(segment);
     }
+    if (segment instanceof MutableSegment) {
+      trackNewlyAddedSegment(segment);
+    }
   }
 
   @Override
   public void untrackSegmentForUpsertView(IndexSegment segment) {
     if (_upsertViewManager != null) {
       _upsertViewManager.untrackSegment(segment);
     }
+    if (segment instanceof MutableSegment) {
+      untrackNewlyAddedSegment(segment);
+    }
+  }
+
+  @VisibleForTesting
+  void trackNewlyAddedSegment(IndexSegment segment) {
+    if (_newSegmentTrackingTimeMs > 0) {
+      _newlyAddedSegments.put(segment.getSegmentName(), -1L);
+    }
+  }
+
+  @VisibleForTesting
+  void untrackNewlyAddedSegment(IndexSegment segment) {
```

Comment:
Why do we need to differentiate track/untrack? It is quite confusing that the segments are actually tracked after calling untrack. Some javadoc can be helpful

Reply:
I left comment for `_newlyAddedSegments` variable, that the delay timer starts after the segment is fully processed, as the time to add a segment can vary a lot.

and similar comment added for the new config 
```
@JsonPropertyDescription("How long to track a newly added segment, after it is processed by the server.")
  private long _newSegmentTrackingTimeMs = 30000;
```

lemme know if these make sense.

Reply:
I understand why we want to track the segment before processing the segment addition, but why we want to untrack it after it is added? Seems like the intention is to always include the segment when it is being added, then start counting timeout after it is fully added?
IMO that might not be necessary because we use this timeout to give some buffer time to broker to reflect the IS change. Can you directly track the segment with timeout before adding it? Or am I missing something here?

Reply:
Yes, I intended to track the segment when it’s being added, and untrack it so we don’t always add such segments to query’s list of selected segments unnecessarily. The delay when to untrack a segment is the time to wait for broker to add it to routing table. I assumed broker adds the segment to query’s list of selected segments when it’s ONLINE in EV. (But you said IS in your comment. I may have missed some details here. Perhaps when uploading a segment, it’ll make all replica groups unavailable as checked by the strictReplicaGroup routing policy… )

But anyway, as an example, if we upload a segment to upsert table and server takes X mins to add it. Then the segment takes at least X mins becomes ONLINE in EV, and then broker may take another few seconds to include it in routing table and start to add it to query’s list of selected segments. In this case, the server needs to track this segment for X mins + a few seconds in order to add the segment to query’s selected segment list speculatively. I assume X may vary a lot among segments, so I define the delay to be counted after the segment is added, to not worry about when X happen to be too long.

Reply:
I see. Yes, I think broker adds it to routing after it turns ONLINE in EV.
I still feel it is a misleading calling it untrack but actually track it with a timeout. How about calling them `preAddNewSegment()` and `postAddNewSegment()`? This is non-blocking comment

Reply:
As just tested, broker reacts on IS changes, and it tracks newly uploaded segments as 'New' and add them to the list of optional segments for server to process. The newly uploaded segments are added to broker's routing table once IS gets updated, and it doesn't wait for the segment to get ONLINE in EV. 

you're right here
> IMO that might not be necessary because we use this timeout to give some buffer time to broker to reflect the IS change....

I can simplify the logic in this PR to kick off the tracking timer when starting to add the segment, instead of waiting for adding segment to finish.